### PR TITLE
[LSP] Add multi config & vscode multi-root workspace support

### LINKED ISSUE #200: Silently fails on unsupported options
Suddenly the extension stopped highlighting Flow errors with no indication of any problems. I ran Flow from the command line and got the following error:

    .flowconfig:11 Unsupported option specified! (unsafe.enable_getters_and_setters)

Removing the unsupported option fixed the problem. It would be great if the extension could catch this error and display it to the user somehow.

### LINKED ISSUE #253: shouldn't you disable flow bin check when no .flowconfig is present?
Hi there, 
I'm running vscode on Ubuntu 16.04 and I have `flow-for-vscode` extension installed. In `user settings` I have already configured 
`  "flow.useNPMPackagedFlow": true  `  

Question 1: do I still have to specify `flow.pathToFlow` in this case?  didn't I make it clear that I want `flow-for-vscode` to look for `flow-bin` in `${workspaceRoot}/node_modules/.bin/flow`?

Question 2: It becomes very annoying that this extension always complain "can't find flow in PATH. try to npm install flow-bin -g" *even when no .flowconfig* is present for the project. Is this to be expected?

I know I can just install a global `flow-bin` to make it happy,  but I don't want to do that; I prefer each project have their own flow version (just like eslint).

Any suggestions?
thanks

### LINKED ISSUE #282: Error message & logic for choosing flow binary are broken
I was trying to get this plugin to work with custom paths or node_modules paths, and noticed that the logic for resolving the flow binary and then displaying the error is very broken. The plugin should not fall back to every possible location of flow [like this](https://github.com/flowtype/flow-for-vscode/blob/80b5cd1b1036036315b1a88991c043fcf5bbfffe/lib/pkg/flow-base/lib/FlowHelpers.js#L174-L183), and then only log an error message relating to the last thing it tried. It really shouldn't try any fallback options at all, if I bothered to set my own specific path then it's very unlikely that if that path is wrong the default fallback is going to be right.

What makes this worse is the error message just says flow wasn't found on PATH. The info logging doesn't show anything better, so there's no way to tell how the plugin interpreted my custom path to figure out why that might not be working.

I assume it was written this way to try to be helpful and make things "just work", but it makes the plugin so much more frustrating to set up.

(btw my actual problem seemed to be that `workspaceFolder` isn't supported and I was trying to use this with a multi-root workspace. Which is a separate issue, there should be indication of this somewhere in the docs if this plugin doesn't support multi-root workspaces).

### LINKED ISSUE #290: does not work without `node` executable available
(on mac os)

it seems this extension requires a `node` executable being available. is this really necessary? i mean, visual-studio-code works without having a `node` installed, shouldn't this extension also work without having to install a `node` executable?

in my specific use case, i use `nvm` to manage node, and i do not have a `default` node. i activate node-versions when i work on the specific project. as a workaround i can do `nvm use lts/carbon` and then start visual-studio-code from that terminal-window and it will work, but i would prefer not having this restriction.

### LINKED ISSUE #302: Plugin doesn't work in a Monorepo environment
Hello everybody,

First off thank you very much 👍

But i've got a problem working with it in a typical lerna project with packages where only some packages make use of flow and therefore i don't have a root-level ".flowconfig", is there some possibility to allow such an environment to work?

And for instance i make heavy use of `module.name_mapper` which i don't know if it would work from a root-level config.

If you need more info or some demo repository, let me know :)

Thanks in advance!

### PR DESCRIPTION
**Pull request by Mayank1791989**
**Client**

New implementation will handle following cases

- Single workspace with multiple flowconfig (multi-config)
```
# some single workspace project

vscode-root
  └──  project-A                [create flow client for this dir]
        └── src                 (contains js files)
        └── .flowconfig
  └── some-dirA
        └── project-c           [no .flowconfig so no client]
              └── src           (contains js files)
              └── test.js
  └──  some-dir
        └── project-B           [create flow client for this dir]
              └── src           (contains js files)
              └── .flowconfig
              └── test.js
```

- Multiple workspace with single|multiple flowconfig (vscode multi-root workspace)

```
# some multiple workspace project

vscode-workspace-folder-a (all clients below will use this worskspaceFolder settings)
  └──  project-aa               [create flow client for this dir]
        └── src                 (contains js files)
        └── .flowconfig
  └── some-dirA
        └── project-ab          [no .flowconfig so no client]
              └── src           (contains js files)
              └── test.js
  └──  some-dir
        └── project-ac           [create flow client for this dir]
              └── src           (contains js files)
              └── .flowconfig
              └── test.js

vscode-workspace-folder-b (all clients below will use this worskspaceFolder settings)
  └──  project-ba               [create flow client for this dir]
        └── src                 (contains js files)
        └── .flowconfig
  └── some-x-dir
        └── project-b           [no .flowconfig so no client]
              └── src           (contains js files)
              └── test.js
  └──  some-y-dir
        └── project-c           [create flow client for this dir]
              └── src           (contains js files)
              └── .flowconfig
              └── test.js
```

- The clients are created lazily when file (javascript or .flowconfig) is opened and .flowconfig is found
for it. (fixes #253)

(fixes #302, #308, #253, #228, #201, #195, #189, #184, #165, #81)

**Resolve Flow-Bin**
- Improve|fix logic to find flow-bin (fixes #282, #240, #209)
- Fix flow bundled with plugin not working
  issue: we are using .bin/flow but vscode never includes .bin in plugin package (see [issue](https://github.com/Microsoft/vscode/issues/53916#issuecomment-404109902))
- Add logs while finding flow-bin to debug issues
- Remove node check (fixes #290)
- `flow.pathToFlow`: value is normalized and .cmd is added if required so same value can be used for linux, osx & win.

**Error handling**  (fixes #200)
- Improve error handling.
- Show action in errors so that user can recover from error without restarting vscode

**Commands**
- Add commands
  - Toggle coverage
  - Show client status
  - Restart Client
  - Log Client Debug Info
  - Show output channel

**Status Bar Widget**

Include more info in status widget to help user know what this plugin is using
- Include flow version
- Add flow info section in widget tooltip which includes
  - path to .flowconfig
  - flow version
  - path to "flow" binary

**Vscode settings**
- `flow.coverageSeverity` to control type coverage diagnostic severity
- `flow.useBundledFlow` enable|disable use of flow bundled with this plugin
- `flow.lazyMode`: to add flow [lazyMode](https://flow.org/en/docs/lang/lazy-modes) support
- `flow.trace.server` log communication between vscode and flow lsp
- `flow.logLevel` control log level of plugin logs

**Docs**
- Update readme

**Build | Tools**
- Change babel target node to 8 (previously we are compiling for node 6)
- Update flow (bundled) to 0.93
- Update vscode-languageclient to 5.2
- Update vscode requirement to match vscode-languageclient requirement.
- Add linting (eslint + prettier)

**Other Changes**
- move non lsp code into directory

### CODE REVIEW

**Comment by thymikee** [comment]
Whoa this is so cool! I need to find some time to take a look at it. @orta have you seen this?

**Comment by jfbrazeau** [comment]
Is it possible to know if this PR will be merged in the coming days ? Not being able to handle multi root workspaces in VSCode is a strong limitation for somebody (like me) who plans to use Flow. As far as I can see, it is even probably a strong limitation for using flow itself as most people use VSCode today.

Il like the "flow" approach about static type checking. I would really like to give it a chance, but I must admit that not being able to use it in VSCode with a multi root project could be a justification to switch to Typescript (at least it doesn't help me to argue against colleagues at work).

That's why I would really appreciate if `flow-for-vscode` team could give a feedback about this PR. Thank you in advance.

**Comment by orta** [comment]
Yeah, I missed this - dang. That's a lot of work! 

Ok, looking now

**Comment by orta** [comment]
Alright, so nothing stands out in the code - and I've given it 10m with the dev build inside the Jest codebase just clicking around and nothing seems to be broken. So I'm game to ship this, nice work @Mayank1791989

**Comment by orta** [comment]
Cool, so this is now shipped as `1.1.0` and should be appearing inside vscode pretty soon for folks

**Comment by jfbrazeau** [comment]
Great news ! Thank you very much !

**Comment by kuus** [comment]
thanks @Mayank1791989!
@orta do you know if this includes support for yarn workspaces?
For instance I have this folder/project structure
```
/project
/project/.flowconfig
/project/node_modules
/project/packages
/project/packages/core
/project/packages/core/.flowconfig
/project/packages/www
/project/packages/www/.flowconfig
```
my `project/package.json` has the following
```
  "workspaces": {
    "nohoist": [
      "gatsby*"
    ],
    "packages": [
      "packages/*"
    ]
  }
```
so when I run yarn install from the two packages the modules are actually put in `/project/node_modules`
instead of in each `/package/package_name` folder.

In VScode I still get a bunch of errors such `Cannot resolve module '@material-ui/core/styles'.` from `/project/packages/core/MyComponent.js`

I am not sure how to setup the `.flowconfig` files in the root and in each package in order to make flow resolve the imports, I've tried solutions such using:
```
[options]
module.system.node.resolve_dirname='<PROJECT_ROOT>../../node_modules'
# or
module.system.node.resolve_dirname=../../node_modules
# or
module.system.node.resolve_dirname=./node_modules
```
and many others but nothing seem to work...

To recap, this is the three `.flowconfig` files I am trying to figure out:

`/project/.flowconfig`:
```
[options]
```

`/project/packages/core/.flowconfig`:
```
[options]
```

`/project/packages/www/.flowconfig`:
```
[options]
```
hope you can help me understand this, thanks in advance!

**Comment by Mayank1791989** [comment]
@kuus The issue you are facing is the limitation of flow, there is nothing we can do in this plugin. 

Have you tried [flow-mono-cli](https://github.com/ImmoweltGroup/flow-mono-cli) It basically creates symlinks of node_modules which are hoisted to top level folder (and as flow supports symlinks it will fix your missing modules errors).

**Comment by Mayank1791989** [comment]
@thymikee and @orta thanks for reviewing and merging the PR.

**Comment by fabiomcosta** [comment]
Is there a case where using `lazyMode` is not recommended?
I guess my question is: shouldn't `lazyMode` be set to `ide` by default?
